hbase-client/src/main/java/org/apache/hadoop/hbase/client/ZKConnectionRegistry.java

```diff
@@ -69,39 +73,85 @@ class ZKConnectionRegistry implements ConnectionRegistry {
   private final ReadOnlyZKClient zk;
 
   private final ZNodePaths znodePaths;
+  private final Configuration conf;
+
+  public static final String EXPECTED_TIMEOUT = "expected.timeout";
+  public static final int DEFAULT_EXPECTED_TIMEOUT = 200000;
+  public static final String MAX_ATTEMPTS = "max.attempts";
+  public static final int DEFAULT_MAX_ATTEMPTS = 5;
+  public static final String PAUSE_NS = "pause.ns";
+  public static final long DEFAULT_PAUSE_NS = 100000;
+
 
   // User not used, but for rpc based registry we need it
   ZKConnectionRegistry(Configuration conf, User ignored) {
     this.znodePaths = new ZNodePaths(conf);
     this.zk = new ReadOnlyZKClient(conf);
+    this.conf = conf;
     if (NEEDS_LOG_WARN) {
       synchronized (WARN_LOCK) {
         if (NEEDS_LOG_WARN) {
           LOG.warn(
             "ZKConnectionRegistry is deprecated. See https://hbase.apache.org/book.html#client.rpcconnectionregistry");
           NEEDS_LOG_WARN = false;
+
         }
       }
     }
   }
 
+  public ZNodePaths getZNodePaths() {
+    return znodePaths;
+  }
+
   private interface Converter<T> {
     T convert(byte[] data) throws Exception;
   }
 
   private <T> CompletableFuture<T> getAndConvert(String path, Converter<T> converter) {
     CompletableFuture<T> future = new CompletableFuture<>();
-    addListener(zk.get(path), (data, error) -> {
-      if (error != null) {
-        future.completeExceptionally(error);
-        return;
-      }
-      try {
-        future.complete(converter.convert(data));
-      } catch (Exception e) {
-        future.completeExceptionally(e);
-      }
-    });
+      TimerTask pollingTask = new TimerTask() {
```

Comment:
I do not think we need to do this here?

We have timeout in ReadOnlyZKClient, here we just need to schedule retries?